[UnifiedPDF] Implement snapshotting

hortont424 opened the pull request:
#### e810cde0dcadb06b50af157ea3486eb36bfa7d46
<pre>
[UnifiedPDF] Implement snapshotting
<a href="https://bugs.webkit.org/show_bug.cgi?id=265630">https://bugs.webkit.org/show_bug.cgi?id=265630</a>
&lt;<a href="https://rdar.apple.com/problem/118550690">rdar://problem/118550690</a>&gt;

Reviewed by Richard Robinson.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::shouldCreateTransientPaintingSnapshot const):
(WebKit::PDFPluginBase::snapshot):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paint):
(WebKit::UnifiedPDFPlugin::snapshot): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::shouldCreateTransientPaintingSnapshot const):
Implement paint() in terms of paintContents() and disable the unnecessary
indirection of PDFPlugin&apos;s ShareableBitmap snapshotting for UnifiedPDFPlugin.

paint() is only called during snapshotting.

Canonical link: <a href="https://commits.webkit.org/271403@main">https://commits.webkit.org/271403@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45edd219adf0b58a662dee678ef838879da54474

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28187 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30716 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25681 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4211 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28455 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24253 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4854 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4989 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31405 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25794 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25680 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31301 "Found 1 new test failure: fast/events/monotonic-event-time.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4975 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3151 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29058 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6523 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6776 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5412 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5490 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -97,12 +97,12 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     id accessibilityObject() const override;
     id accessibilityAssociatedPluginParentForElement(WebCore::Element*) const override;
 
+    void paint(WebCore::GraphicsContext&, const WebCore::IntRect&) override;
```
Whoops, should put this up with its friends.

aprotyas replied:
> Whoops, should put this up with its friends.

Agreed.

aprotyas replied:
Also, this can be a const function, right?

aprotyas replied:
Never mind, it can't because of the `-[PDFPage drawWithBox:toContext:]` call in `UnifiedPDFPlugin::paintContents`.

hortont424 replied:
it also comes all the way from Widget and I have no interest in constifying that whole tree :)

hortont424 replied:
Also turns out it has no friends (I was thinking of the base header)

aprotyas reviewed (commented):
LGTM

hortont424 committed:
[UnifiedPDF] Implement snapshotting
https://bugs.webkit.org/show_bug.cgi?id=265630
<rdar://problem/118550690>

Reviewed by Richard Robinson.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::shouldCreateTransientPaintingSnapshot const):
(WebKit::PDFPluginBase::snapshot):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paint):
(WebKit::UnifiedPDFPlugin::snapshot): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::shouldCreateTransientPaintingSnapshot const):
Implement paint() in terms of paintContents() and disable the unnecessary
indirection of PDFPlugin's ShareableBitmap snapshotting for UnifiedPDFPlugin.

paint() is only called during snapshotting.

Canonical link: https://commits.webkit.org/271403@main
